python/ray/serve/pipeline/deployment_function_node.py

```diff
@@ -30,19 +32,32 @@ def __init__(
             other_args_to_resolve=other_args_to_resolve,
         )
 
-        if "deployment_self" in self._bound_other_args_to_resolve:
-            original_deployment: Deployment = self._bound_other_args_to_resolve[
-                "deployment_self"
+        if "deployment_schema" in self._bound_other_args_to_resolve:
+            deployment_schema: DeploymentSchema = self._bound_other_args_to_resolve[
+                "deployment_schema"
             ]
-            self._deployment = original_deployment.options(
-                name=(
-                    deployment_name
-                    if original_deployment._name
-                    == original_deployment.func_or_class.__name__
-                    else original_deployment._name
-                ),
-                init_args=tuple(),
+            deployment_shell = schema_to_deployment(deployment_schema)
+
+            # Prefer user specified name to override the generated one.
+            if (
+                inspect.isfunction(func_body)
+                and deployment_shell._name != func_body.__name__
+            ):
+                self._deployment_name = deployment_shell._name
+
+            # Set the route prefix, prefer the one user supplied,
+            # otherwise set it to /deployment_name
+            if deployment_shell._route_prefix != f"/{deployment_shell._name}":
+                route_prefix = deployment_shell._route_prefix
+            else:
+                route_prefix = f"/{deployment_name}"
+
+            self._deployment = deployment_shell.options(
+                func_or_class=func_body,
+                name=self._deployment_name,
+                init_args=(),
                 init_kwargs=dict(),
+                route_prefix=route_prefix,
             )
         else:
```

Comment:
i think this else condition exists to take care of the "popped deployment_self" path and it was hit in `test_pipeline_dag.py` after the first `to_json` call to deserialize the `DeploymentNode`. With new changes included in `other_args_to_resolve` I don't think we're hitting this anymore ?

There're a few places in `DeploymentNode`, `DeploymentMethodNode` and `DeploymentFunctionNode` that we're creating `Deployment` object with `DeploymentConfig()`, with this change we should remove them all to avoid confusion

Reply:
you are still hitting these condition in test_generate in the serve/pipeline codebase

Reply:
hmm i see so in serve path we only hit the if now given we don't pop the key anymore, and all else goes to ray_dag -> serve_dag path. i good with it as it then, we can address later.